File: test_utils/test_mixin.py
import json
import os
from pathlib import Path
from typing import Dict, Optional, TypedDict

class TestConfig(TypedDict):
    type: str
    framework: str
    service: str

class TestConfigFile(TypedDict):
    testFiles: Dict[str, TestConfig]

class TestMixin:
    def __init__(self, current_file: str):
        self.current_file = current_file
        self.index_path = Path(os.getcwd()) / 'test-index.json'
        self.config = self._load_config()

    def _load_config(self) -> TestConfigFile:
        with open(self.index_path, 'r') as f:
            return json.load(f)

    def _save_config(self) -> None:
        with open(self.index_path, 'w') as f:
            json.dump(self.config, f, indent=2)

    def get_test_config(self) -> Optional[TestConfig]:
        # Try exact match first
        if self.current_file in self.config['testFiles']:
            return self.config['testFiles'][self.current_file]

        # Try relative path match
        relative_path = os.path.relpath(self.current_file, os.getcwd())
        if relative_path in self.config['testFiles']:
            return self.config['testFiles'][relative_path]

        return None

    def get_service_name(self) -> str:
        config = self.get_test_config()
        return config['service'] if config else 'unknown'

    def get_test_type(self) -> str:
        config = self.get_test_config()
        return config['type'] if config else 'unknown'

    def get_framework(self) -> str:
        config = self.get_test_config()
        return config['framework'] if config else 'unknown'

    def get_test_metadata(self) -> str:
        config = self.get_test_config()
        if not config:
            return 'Unknown test configuration'
        return f"{config['type']} test using {config['framework']} framework for {config['service']} service"

    def index_test(self, command: str) -> None:
        relative_path = os.path.relpath(self.current_file, os.getcwd())
        test_type = self.get_test_type()
        framework = self.get_framework()
        service = self.get_service_name()

        # Update the test index
        self.config['testFiles'][relative_path] = {
            'type': test_type,
            'framework': framework,
            'service': service
        }

        # Save the updated index
        self._save_config()

        # Log the command for reference
        print(f"Test command for {relative_path}:")
        print(command)

def pytestMixin(file_path: str):
    """Create a pytest mixin for the given file."""
    mixin = TestMixin(file_path)
    return {
        'setup_module': lambda: print(f"Running {mixin.get_test_metadata()}"),
        'teardown_module': lambda: print(f"Completed {mixin.get_test_metadata()}")
    } 

File: test-utils/test-mixin.ts
import * as fs from 'fs';
import * as path from 'path';

interface TestConfig {
  type: 'pytest' | 'jest';
  framework: 'pact' | 'unit' | 'integration';
  service: string;
}

interface TestConfigFile {
  testFiles: {
    [key: string]: TestConfig;
  };
}

export class TestMixin {
  private config: TestConfigFile;
  private currentFile: string;
  private indexPath: string;

  constructor(currentFilePath: string) {
    this.currentFile = currentFilePath;
    this.indexPath = path.resolve(process.cwd(), 'test-index.json');
    this.config = this.loadConfig();
  }

  private loadConfig(): TestConfigFile {
    const configContent = fs.readFileSync(this.indexPath, 'utf-8');
    return JSON.parse(configContent);
  }

  private saveConfig(): void {
    fs.writeFileSync(this.indexPath, JSON.stringify(this.config, null, 2));
  }

  public getTestConfig(): TestConfig | null {
    // Try exact match first
    if (this.config.testFiles[this.currentFile]) {
      return this.config.testFiles[this.currentFile];
    }

    // Try relative path match
    const relativePath = path.relative(process.cwd(), this.currentFile);
    if (this.config.testFiles[relativePath]) {
      return this.config.testFiles[relativePath];
    }

    return null;
  }

  public getServiceName(): string {
    const config = this.getTestConfig();
    return config?.service || 'unknown';
  }

  public getTestType(): string {
    const config = this.getTestConfig();
    return config?.type || 'unknown';
  }

  public getFramework(): string {
    const config = this.getTestConfig();
    return config?.framework || 'unknown';
  }

  public getTestMetadata(): string {
    const config = this.getTestConfig();
    if (!config) {
      return 'Unknown test configuration';
    }
    return `${config.type} test using ${config.framework} framework for ${config.service} service`;
  }

  public async indexTest(command: string): Promise<void> {
    const relativePath = path.relative(process.cwd(), this.currentFile);
    const testType = this.getTestType();
    const framework = this.getFramework();
    const service = this.getServiceName();

    // Update the test index
    this.config.testFiles[relativePath] = {
      type: testType as 'pytest' | 'jest',
      framework: framework as 'pact' | 'unit' | 'integration',
      service
    };

    // Save the updated index
    this.saveConfig();

    // Log the command for reference
    console.log(`Test command for ${relativePath}:`);
    console.log(command);
  }
}

// Example usage for Jest:
export const withTestMixin = (filePath: string) => {
  const mixin = new TestMixin(filePath);
  return {
    beforeAll: (callback: () => void) => {
      console.log(`Running ${mixin.getTestMetadata()}`);
      callback();
    },
    afterAll: (callback: () => void) => {
      console.log(`Completed ${mixin.getTestMetadata()}`);
      callback();
    }
  };
};

// Example usage for Pytest:
export const pytestMixin = (filePath: string) => {
  const mixin = new TestMixin(filePath);
  return {
    setup_module: () => {
      console.log(`Running ${mixin.getTestMetadata()}`);
    },
    teardown_module: () => {
      console.log(`Completed ${mixin.getTestMetadata()}`);
    }
  };
}; 